File: samples/src/main/java/jvmMain/java/legacy/LegacyTransferToUser.java
package jvmMain.java.legacy;

import jvmMain.java.Common;
import jvmMain.java.Config;
import one.mixin.bot.SessionKt;
import one.mixin.bot.api.MixinResponse;
import one.mixin.bot.extension.ByteArrayExtensionKt;
import one.mixin.bot.safe.TipBody;
import one.mixin.bot.util.CryptoUtilKt;
import one.mixin.bot.vo.Snapshot;
import one.mixin.bot.vo.TransferRequest;

import java.io.IOException;
import java.util.UUID;

public class LegacyTransferToUser {
    public static void main(String[] args) throws IOException {

        byte[] serverPublicKey = ByteArrayExtensionKt.hexStringToByteArray(Config.BOT_SERVER_PUBLIC_KEY);
        byte[] sessionPrivateKey = ByteArrayExtensionKt.hexStringToByteArray(Config.BOT_SESSION_PRIVATE_KEY);
        byte[] pinToken = CryptoUtilKt.decryptPinToken(CryptoUtilKt.publicKeyToCurve25519(serverPublicKey), sessionPrivateKey);

        final String assetId = Common.Token.CNB.getAssetId();
        final String amount = "0.00000001";
        final String traceId = UUID.randomUUID().toString();
        final String memo = "Hello mixin java";
        final String userId = "cfb018b0-eaf7-40ec-9e07-28a5158f1269";

        byte[] signTarget = TipBody.forTransfer(assetId, userId, amount, traceId, memo);
        String pin = SessionKt.encryptTipPin(pinToken, signTarget, ByteArrayExtensionKt.hexStringToByteArray(Config.BOT_SPEND_KEY));

        MixinResponse<Snapshot> response = Common.botClient.getSnapshotService().transferCall(new TransferRequest(
                assetId, userId, amount, pin, traceId, memo, null
        )).execute().body();
        System.out.println(response);
    }

}

File: samples/src/main/java/jvmMain/kotlin/legacy/LegacyTransferToUser.kt
package jvmMain.kotlin.legacy

import jvmMain.kotlin.Config
import jvmMain.kotlin.Token
import jvmMain.kotlin.botClient
import kotlinx.coroutines.runBlocking
import one.mixin.bot.encryptTipPin
import one.mixin.bot.extension.hexStringToByteArray
import one.mixin.bot.safe.TipBody
import one.mixin.bot.util.decryptPinToken
import one.mixin.bot.util.publicKeyToCurve25519
import one.mixin.bot.vo.TransferRequest
import java.util.UUID


fun main(): Unit = runBlocking {
    val pinToken = publicKeyToCurve25519(Config.BOT_SERVER_PUBLIC_KEY.hexStringToByteArray())
    val aesKey = decryptPinToken(
        pinToken,
        Config.BOT_SESSION_PRIVATE_KEY.hexStringToByteArray(),
    )

    val assetId = Token.CNB.assetId
    val amount = "0.0000001"
    val userId = "cfb018b0-eaf7-40ec-9e07-28a5158f1269"
    val traceId = UUID.randomUUID().toString()
    val memo = "test"

    val signTarget = TipBody.forTransfer(
        assetId = assetId,
        amount = amount,
        counterUserId = userId,
        traceId = traceId,
        memo = memo,
    )
    val pin = encryptTipPin(
        aesKey,
        signTarget,
        Config.BOT_SPEND_KEY.hexStringToByteArray(),
    )
    botClient.snapshotService.transfer(
        TransferRequest(
            assetId = assetId,
            amount = amount,
            opponentId = userId,
            traceId = traceId,
            memo = memo,
            pin = pin,
        )
    )
}